src/Agent.PluginCore/TaskPlugin.cs

```diff
@@ -0,0 +1,224 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.OAuth;
+using Microsoft.VisualStudio.Services.WebApi;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Pipelines = Microsoft.TeamFoundation.DistributedTask.Pipelines;
+
+namespace Agent.PluginCore
+{
+    public interface IAgentTaskPlugin
+    {
+        Guid Id { get; }
+        string Version { get; }
+        string FriendlyName { get; }
+        string Description { get; }
+        string HelpMarkDown { get; }
+        string Author { get; }
+        TaskInputDefinition[] Inputs { get; }
+        HashSet<string> Stages { get; }
+        Task RunAsync(AgentTaskPluginExecutionContext executionContext, CancellationToken token);
+    }
+
+    public class AgentTaskPluginExecutionContext
+    {
+        public AgentTaskPluginExecutionContext()
+        {
+            this.Endpoints = new List<ServiceEndpoint>();
+            this.Inputs = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            this.Repositories = new List<Pipelines.RepositoryResource>();
+            this.TaskVariables = new Dictionary<string, VariableValue>(StringComparer.OrdinalIgnoreCase);
+            this.Variables = new Dictionary<string, VariableValue>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        public string Stage { get; set; }
+        public List<ServiceEndpoint> Endpoints { get; set; }
+        public List<Pipelines.RepositoryResource> Repositories { get; set; }
+        public Dictionary<string, VariableValue> Variables { get; set; }
+        public Dictionary<string, VariableValue> TaskVariables { get; set; }
+        public Dictionary<string, string> Inputs { get; set; }
+
+        public string GetInput(string name, bool required = false)
```

Comment:
we need a vault concept. we can port from powershell, but i'm not sure if it will work on linux.

Reply:
i forgot that. it should work, we can just use AES.